internal/ebpf/probesfacade/captor/capture.go

```diff
@@ -107,51 +119,110 @@ func Capture(functionSymbol string, cmdArgs []string, opts CaptureOptions) ([]ui
+		eventsCh: eventsChannel,
+		lostCh:   lostChannel,
+		opts:     opts,
+		cmd:      cmdArgs,
+	}, nil
+}
+
+// Close closes the ebpf link and module.
+func (ebpf *ebpfSetup) Close() {
+	ebpf.link.Destroy()
+	ebpf.mod.Close()
+}
+
+// Capture collects syscalls from the executed command.
+// Returns values through the given channels.
+func (ebpf *ebpfSetup) Capture(collectedSyscalls chan []uint32, errorCh chan error) {
+	// setting up ticker to dump results
+	// every interval of time.
+	var ticker *time.Ticker
+	var interval time.Duration
+	if ebpf.opts.Interval > 0 {
+		interval = time.Duration(ebpf.opts.Interval) * time.Second
+	} else {
+		// wait what? ~114 years of interval?
+		// yes, so that the ticker will never be triggered.
+		// the value looks reasonably good.
+		interval = time.Duration(1000000) * time.Hour
+	}
+	ticker = time.NewTicker(interval)
+	defer ticker.Stop()
+
 	var wg sync.WaitGroup
 	wg.Add(1)
-	outputCh := make(chan string)
-	errorCh := make(chan string)
-	go executor.Run(cmdArgs, opts.CommandOutput, opts.CommandError, &wg, outputCh, errorCh)
+	cmdStdoutCh := make(chan string)
+	cmdStderrCh := make(chan string)
+	// running command to trace its syscalls
+	go executor.Run(
+		ebpf.cmd,
+		ebpf.opts.CommandOutput,
+		ebpf.opts.CommandError,
+		&wg,
+		cmdStdoutCh,
+		cmdStderrCh,
+	)
 
 	var syscalls []uint32
 	go func() {
 		for {
 			select {
-			case data := <-eventsChannel:
+			case data := <-ebpf.eventsCh:
 				var e event
 				err := binary.Read(bytes.NewBuffer(data), binary.LittleEndian, &e)
 				if err != nil {
-					fmt.Fprintf(os.Stderr, "error reading event: %v\n", err)
+					//fmt.Fprintf(os.Stderr, "error reading event: %v\n", err)
+					// will be left empty for now.
 					return
 				}
 				syscalls = append(syscalls, e.SyscallID)
-			case lost := <-lostChannel:
-				fmt.Fprintf(os.Stderr, "lost %d data\n", lost)
+			case <-ticker.C:
+				// used to send incremental result
+				// every interval of time.
+				collectedSyscalls <- syscalls
+				// we clear the slice after sending data
+				// to the channel, so on the next iteration
+				// this will have only the most recent values.
+				syscalls = nil
+			case _ = <-ebpf.lostCh:
+				// managing errors from libbpf
+				// will be left empty for now.
+				//fmt.Fprintf(os.Stderr, "lost %d data\n", lost)
 				return
-			case line, ok := <-outputCh:
+			case line, ok := <-cmdStdoutCh:
+				// managing stdout from executed command
 				if !ok {
 					break
 				}
-				fmt.Println(line)
-			case err, ok := <-errorCh:
+				fmt.Println("stdout:", line)
+			case err, ok := <-cmdStderrCh:
+				// managing stderr from executed command
 				if !ok {
 					break
 				}
-				fmt.Println(err)
+				fmt.Println("stderr:", err)
 			}
 		}
 	}()
 
-	rb.Poll(300)
+	ebpf.pb.Poll(300)
 	// wait for args completion
 	wg.Wait()
-	rb.Stop()
+	ebpf.pb.Stop()
 
-	return syscalls, nil
+	// sending last remained syscalls
+	// and close the channel.
+	collectedSyscalls <- syscalls
```

Comment:
I would close these input chaannels in the outside context
Here we're interfering with the lifecycle of a thing that is born outside of this context

I would move it to a defer after the make of these very channels

Reply:
I agree with you, but I see one problem. If the channels are not closed, the `for {}` loop in `cmd/capture.go` can't be stopped. so closing the channels from `internal/ebpf/probesfacade/captor/capture.go` was done for this reason.

Reply:
maybe then handling a signal that tells the main context that you're done?

Reply:
not sure I got it. can you provide an example?

Reply:
for example, encode an "end channel" value (dunno, MAX_INT for example) and the main process can interpret it to stop the loop, close the channel

Reply:
> for example, encode an "end channel" value (dunno, MAX_INT for example) and the main process can interpret it to stop the loop, close the channel
Honestly I don't like the idea that much. What do you think @maxgio92 ?

Reply:
@alegrey91 I agree with @alegrey91. I would close it in the same goroutine where you create it. Theoretically, you could have other consumers of that channel.

Reply:
@maxgio92 @89luca89 I moved the channel closure outside the `Capture` function

Reply:
daje